types/validator_set.go

```diff
@@ -270,34 +270,27 @@ func (vals *ValidatorSet) VerifyCommit(chainID string, blockID BlockID, height i
 	}
 
 	talliedVotingPower := int64(0)
-	round := commit.Round()
+
+	baseVote := Vote{
+		Height:  height,
+		Round:   commit.Round(),
+		Type:    VoteTypePrecommit,
+		BlockID: blockID,
+	}
 
 	for idx, precommit := range commit.Precommits {
 		if precommit == nil {
 			continue // OK, some precommits can be missing.
 		}
-		if precommit.Height != height {
-			return fmt.Errorf("Invalid commit -- wrong height: want %v got %v", height, precommit.Height)
-		}
-		if precommit.Round != round {
-			return fmt.Errorf("Invalid commit -- wrong round: want %v got %v", round, precommit.Round)
-		}
-		if precommit.Type != VoteTypePrecommit {
-			return fmt.Errorf("Invalid commit -- not precommit @ index %v", idx)
-		}
 		_, val := vals.GetByIndex(idx)
+		baseVote.Timestamp = precommit.Timestamp
 		// Validate signature.
-		precommitSignBytes := precommit.SignBytes(chainID)
+		precommitSignBytes := baseVote.SignBytes(chainID)
 		if !val.PubKey.VerifyBytes(precommitSignBytes, precommit.Signature) {
 			return fmt.Errorf("Invalid commit -- invalid signature: %v", precommit)
 		}
 		// Good precommit!
-		if blockID.Equals(precommit.BlockID) {
-			talliedVotingPower += val.VotingPower
-		} else {
-			// It's OK that the BlockID doesn't match.  We include stray
-			// precommits to measure validator availability.
```

Comment:
@jaekwon just pointed out we might need more here, since currently we wouldn't be including votes for validators that correctly precommited for nil. 

We could rectify that by including precommits for nil (note this is distinct from nil precommits, ie. not receiving a precommit for anything from that validator), and including a bit array indicating which votes are for nil and which are for the block id

Reply:
I think we should include nil precommits, since precluding those would have adverse economic consequences and could lead to unexpected incentive distortions.